[BE-Fix] 캘린더 연동 시 발생하는 버그 수정

kwon204 opened the pull request:
<!-- 제목 템플릿
[파트명-Feat] [파트명-Refactor] [파트명-Fix] [파트명-Chore] [파트명-Remove]
-->
## #️⃣ 연관된 이슈>
- 캘린더 연동 후 구글 서버로부터 가져온 일정들이 DB에 저장할 때 userService.getCurrentUser()를 시도하여 에러가 발생함.
- 클라이언트 요청에 의해 일정 생성 로직과 동기화를 위해 일정 생성하는 로직을 분리하여 해결

## 수정 사항 (25.07.13)
- 캘린더 연동 요청 후 캘린더 정보 저장까지 비동기에서 동기로 처리하도록 수정
- TransactionalEventListener의 BEFORE_COMMIT 옵션에서 문제가 있어 코드를 수정하였습니다. 
  - 이벤트를 발행하는 메소드가 커밋되기 전에 실행되고 이벤트 리스너 쪽의 코드와 함께 커밋되는 것은 맞으나, 구글 토큰을 저장하기 전에 토큰을 사용하려고 시도하는 경우가 발생할 수 있기 때문에 문제가 됩니다. 
  - 따라서 **BEFORE_COMMIT을 사용할 수 없습니다.** 
- 그리고 서치를 하던 중, 트랜잭션을 물고 외부 API를 호출하게 되면 결과를 기다리는 동안 트랜잭션을 물고 있어 자원을 낭비하는 코드 구조라고 합니다. 그래서 기존의 linkGoogleCalendar에 적용했던 **트랜잭션을 제거하는 것**이 필요하였습니다. 
- 그리고 프론트에서 필요한 것은 캘린더의 정보만 필요하기 때문에 캘린더의 일정을 모두 가져오는 syncWithCalendar는 동기로 처리할 필요가 없다고 생각하였습니다.  그래서 **syncWithCalendar는 비동기로 처리**하는 게 적절하다고 생각하였습니다. 

---
- 위 내용을 종합해서 Facade 패턴을 사용해서 구현하였습니다. 
- 기존의 코드에서 조금만 변경해서 이벤트를 연속해서 발행하는 것도 생각해봤는데, 너무 복잡해지는 문제가 있었습니다. 
- CalendarFacade의 linkGoogleCalendar는 트랜잭션을 적용하지 않고 googleCalendarService를 직접 호출해서 getPrimaryCalendar는 순차적으로 진행합니다. 그리고 비동기 처리가 가능한 syncWithCalendar는 이벤트를 발행하여 기존과 동일하게 진행됩니다. 


## 📝 작업 내용> 이번 PR에서 작업한 내용을 간략히 설명해주세요(이미지 첨부 가능)

## 🙏 여기는 꼭 봐주세요! > 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **Refactor**
  - Improved internal handling of personal event creation for better maintainability and consistency.
  - Introduced a new calendar management service to streamline Google Calendar linking and retrieval.
  - Updated calendar and user services to support enhanced token management and calendar operations.
  - Adjusted event handling for Google Calendar integration to optimize timing and subscription behavior.

- **Tests**
  - Updated tests to align with the improved personal event creation logic and response format.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kwon204 committed:
fix: 캘린더 연동 시 발생하는 버그 수정

- 캘린더 연동 후 구글 서버로부터 가져온 일정들이 DB에 저장할 때 userService.getCurrentUser()를 시도하여 에러가 발생함.
- 클라이언트 요청에 의해 일정 생성 로직과 동기화를 위해 일정 생성하는 로직을 분리하여 해결

kwon204 committed:
fix: 캘린더 연동 후 캘린더 정보가 비어있는 버그 수정

- 캘린더 연동 요청 후 캘린더 정보 저장까지 동기로 처리하도록 수정

efdao commented on backend/src/main/java/endolphin/backend/global/google/event/handler/GoogleEventHandler.java:
before commit으로 하신 이유가 있을까여?

efdao replied:
잘 몰라서 그냥 질문임당

kwon204 replied:
기존에 비동기로 처리할 때는 AFTER_COMMIT이어도 다른 쓰레드로 실행되어 새로운 트랜잭션을 시작하게 됩니다. 하지만 동기로 처리할 때 AFTER_COMMIT인 상태라면 이전의 트랜잭션 내에서 실행하게 되는데, 이 때 commit이 실행된 이후이기 때문에 이후에 실행되는 SQL이 commit되지 않는 문제가 있다고 합니다. 그래서 BEFORE_COMMIT으로 하게 되었습니다. 

그런데 더 찾아보니 현재 구조에서 BEFORE_COMMIT을 하면 문제가 생길 거 같아서 수정하고 다시 올리겠습니다

efdao reviewed (approved):
👍

hamo-o commented:
머지 기원 1일차

kwon204 committed:
refactor: google token 저장 로직을 userService로 이동

kwon204 committed:
refactor: CalendarFacade 추가

- linkGoogleCalendar, getCalendarList를 CalendarFacade로 이동
- CalendarService에서 linkGoogleCalendar, getCalendarList를 삭제
- CalendarController는 CalendarFacade 사용

kwon204 committed:
refactor: 사용하지 않는 메소드 upsertGoogleCalendar 삭제

- @TransactionalEventListener를 @EventListener로 변경
- @Async 추가

kwon204 committed:
chore: 사용하지 않는 테스트 코드 삭제
